src/backend/distributed/metadata/dependency.c

```diff
@@ -742,60 +743,83 @@ SupportedDependencyByCitus(const ObjectAddress *address)
 
 
 /*
- * EnsureRelationDependenciesCanBeDistributed ensures all dependencies of the relation
- * can be distributed.
+ * DeferErrorIfHasUnsupportedDependency returns deferred error message if the given
+ * object has any undistributable dependency.
  */
-void
-EnsureRelationDependenciesCanBeDistributed(ObjectAddress *relationAddress)
+DeferredErrorMessage *
+DeferErrorIfHasUnsupportedDependency(const ObjectAddress *objectAddress)
 {
-	ObjectAddress *undistributableDependency =
-		GetUndistributableDependency(relationAddress);
+	ObjectAddress *undistributableDependency = GetUndistributableDependency(
+		objectAddress);
 
-	if (undistributableDependency != NULL)
+	if (undistributableDependency == NULL)
 	{
-		char *tableName = get_rel_name(relationAddress->objectId);
+		return NULL;
+	}
 
-		if (SupportedDependencyByCitus(undistributableDependency))
-		{
-			/*
-			 * Citus can't distribute some relations as dependency, although those
-			 * types as supported by Citus. So we can use get_rel_name directly
-			 *
-			 * For now the relations are the only type that is supported by Citus
-			 * but can not be distributed as dependency, though we've added an
-			 * explicit check below as well to not to break the logic here in case
-			 * GetUndistributableDependency changes.
-			 */
-			if (getObjectClass(undistributableDependency) == OCLASS_CLASS)
-			{
-				char *dependentRelationName = get_rel_name(
-					undistributableDependency->objectId);
+	char *objectDescription = NULL;
+	char *dependencyDescription = NULL;
+	StringInfo errorMessage = makeStringInfo();
+	StringInfo detailMessage = makeStringInfo();
 
-				ereport(ERROR, (errmsg("Relation \"%s\" has dependency to a table"
-									   " \"%s\" that is not in Citus' metadata",
-									   tableName, dependentRelationName),
-								errhint("Distribute dependent relation first.")));
-			}
-		}
+	#if PG_VERSION_NUM >= PG_VERSION_14
+	objectDescription = getObjectDescription(objectAddress, false);
+	dependencyDescription = getObjectDescription(undistributableDependency, false);
+	#else
+	objectDescription = getObjectDescription(objectAddress);
+	dependencyDescription = getObjectDescription(undistributableDependency);
+	#endif
+
+	/*
+	 * If the given object is a procedure, we want to create it locally,
+	 * so provide that information in the error detail.
+	 */
+	if (getObjectClass(objectAddress) == OCLASS_PROC)
+	{
+		appendStringInfo(detailMessage, "\"%s\" will be created only locally",
+						 objectDescription);
+	}
+	else
+	{
+		detailMessage->data = NULL;
```

Comment:
I don't think we should set this to NULL, just do nothing, isn't it already an empty string? Because that is messing the internal of stringInfo, which might cause problems if it is ever used.


Maybe use as `strlen(detailMessage->data)  == 0 ? NULL : detailMessage->data` to pass NULL to DeferredError?